crates/transaction-generator-lib/src/p2p_transaction_generator.rs

```diff
@@ -13,29 +13,166 @@ use rand::{
+    BurnAndRecycle(usize),
+}
+
+/// Specifies how to get a given number of samples from an item pool.
+pub trait Sampler<T>: Send + Sync {
+    fn sample_from_pool(
+        &mut self,
+        rng: &mut StdRng,
+        pool: &mut Vec<T>,
+        num_samples: usize,
+    ) -> Vec<T>;
+}
+
+/// A sampler that samples a random subset of the pool. Samples are replaced immediately.
+pub struct BasicSampler<T> {
+    phantom: PhantomData<T>,
+}
+
+impl<T> BasicSampler<T> {
+    fn new() -> Self {
+        Self {
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<T: Clone + Send + Sync> Sampler<T> for BasicSampler<T> {
+    fn sample_from_pool(
+        &mut self,
+        rng: &mut StdRng,
+        pool: &mut Vec<T>,
+        num_samples: usize,
+    ) -> Vec<T> {
+        let mut samples = Vec::with_capacity(num_samples);
+        let num_available = pool.len();
+        for _ in 0..num_samples {
+            let idx = rng.gen_range(0, num_available);
+            samples.push(pool[idx].clone());
+        }
+        samples
+    }
+}
+
+/// A samplers that samples from a pool but do not replace items until the pool is depleted.
+/// The pool is divided into sub-pools. Replacement is done with with each sub-pool shuffled internally.
+///
+/// Here is an example. Say the initial pool is `[I, J, K, X, Y, Z]`.
+/// A `BurnAndRecycleSampler` is created with `replace_batch_size=3` to sample from the pool.
+/// The first 6 samples are guaranteed to be `Z`, `Y`, `X`, `K`, `J`, `I`.
+/// Then at the beginning of the 7-th sampling,
+/// sub-pools `{I, J, K}`, `{X, Y, Z}` are shuffled and replaced.
+/// A possible state of the pool is `[K, I, J, Y, X, Z]`.
+///
+/// This behavior helps generate a block of non-conflicting coin transfer transactions,
+/// when there are 2+ sub-pools of size larger than or equal to the block size.
+pub struct BurnAndRecycleSampler<T> {
+    /// We store all sub-pools together in 1 Vec: `item_pool[segment_size * x..segment_size * (x+1)]` being the x-th sub-pool.
+    to_be_replaced: Vec<T>,
+    sub_pool_size: usize,
+}
+
+impl<T: Clone + Send + Sync> BurnAndRecycleSampler<T> {
+    fn new(replace_batch_size: usize) -> Self {
+        Self {
+            to_be_replaced: vec![],
+            sub_pool_size: replace_batch_size,
+        }
+    }
+
+    fn sample_one_from_pool(&mut self, rng: &mut StdRng, pool: &mut Vec<T>) -> T {
+        if pool.is_empty() {
+            let num_addresses = self.to_be_replaced.len();
+            for replace_batch_start in (0..num_addresses).step_by(self.sub_pool_size) {
+                let end = min(replace_batch_start + self.sub_pool_size, num_addresses);
+                self.to_be_replaced[replace_batch_start..end].shuffle(rng);
+            }
+            for _ in 0..num_addresses {
+                pool.push(self.to_be_replaced.pop().unwrap());
+            }
+        }
+        let sample = pool.pop().unwrap();
+        self.to_be_replaced.push(sample.clone());
```

Comment:
having this be a template is nice, but .clone() will not work for the other type (LocalAccount). so if we end up needing it for that, we would need to return &T

But since I'm not sure we will need that, this is fine :)